### PR TITLE
PromQL: Fix unary operator precedence

### LINKED ISSUE #6575: PromQL: Unusual unary operator precedence
Consider the expression:
```
-a^b
```
For written mathematics the convention is to interpret this as follows:
```
-(a^b)
```
However PromQL will interpret it as
```
(-a)^b
```

__To be discussed:__ Should PromQL keep the existing behaviour that might be not expected by users or change it to follow mathematical conventions, at the risk of breaking existing queries?

### PR DESCRIPTION
**Pull request by slrtbtfs**
Fixes #6575

Signed-off-by: Tobias Guggenmos <tguggenm@redhat.com>

### CODE REVIEW

**Comment by brian-brazil** [comment]
Can you add some PromQL tests too?

**Comment by slrtbtfs** [comment]
> Can you add some PromQL tests too?

In the engine?

I've already added a couple in for the parser.

**Comment by brian-brazil** [comment]
In the testdata. Weirdness here is more likely to be spotted at the PromQL level.

**Comment by slrtbtfs** [comment]
Ok, I've added some tests there.

File: promql/testdata/operators.test [hunk 1]
```diff
@@ -35,6 +35,16 @@ eval instant at 50m - - - SUM(http_requests) BY (job)
 eval instant at 50m - - - 1
   -1
 
+eval instant at 50m -2^---1*3
+  -0.75
+
+eval instant at 50m 2/-2^---1*3+2
+  5
+
+eval instant at 50m -10^3 * - SUM(http_requests) BY (job) ^ -1
```
Comment (brian-brazil) [inline promql/testdata/operators.test]: This doesn't seem right, `-x ^ -1` should be same as `-1 / x `

File: promql/testdata/operators.test [hunk 2]
```diff
@@ -35,6 +35,16 @@ eval instant at 50m - - - SUM(http_requests) BY (job)
 eval instant at 50m - - - 1
   -1
 
+eval instant at 50m -2^---1*3
+  -0.75
```
Comment (brian-brazil) [inline promql/testdata/operators.test]: Shouldn't this be -1.5?

Comment (brian-brazil) [inline promql/testdata/operators.test]: Shouldn't thus be -1.5?

File: promql/testdata/operators.test (hunk 1 above)
Comment (slrtbtfs) [reply promql/testdata/operators.test]: It is.

You can get to this expression by transforming the below expression:

```
    1000 / SUM(http_requests) BY (job)
=  - 1000 / - SUM(http_requests) BY (job)            
=  - 10^3 / - SUM(http_requests) BY (job)
=  - 10^3 * -1 / SUM(http_requests) BY (job)        | -1 / x = -x ^ -1
=  - 10^3 * - SUM(http_requests) BY (job) ^ -1
```

File: promql/testdata/operators.test (hunk 2 above)
Comment (slrtbtfs) [reply promql/testdata/operators.test]: Yes, it should.

Comment (slrtbtfs) [reply promql/testdata/operators.test]: This proves that math students are bad at calculating. q.e.d.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**